File: npm-packages/meteor-capacitor/android/src/main/java/com/meteorjs/capacitor/BundleOrganizer.java
/*
 * Vendored from https://github.com/banjerluke/capacitor-meteor-webapp
 * Copyright 2025 Luke Abbott. MIT license. See LICENSES/banjerluke-capacitor-meteor-webapp.txt.
 */

package com.meteorjs.capacitor;

import android.util.Log;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

final class BundleOrganizer {

    private static final String LOG_TAG = "CapacitorMeteorWebApp";

    private static final String SHIM_SCRIPT =
        "<script>\n"
            + "(function() {\n"
            + "    if (window.WebAppLocalServer) return;\n"
            + "\n"
            + "    setupWebAppLocalServer();\n"
            + "\n"
            + "    function setupWebAppLocalServer() {\n"
            + "        var _P;\n"
            + "        function getPlugin() {\n"
            + "            if (!_P) _P = ((window.Capacitor || {}).Plugins || {}).CapacitorMeteorWebApp;\n"
            + "            if (!_P) console.warn('WebAppLocalServer shim: CapacitorMeteorWebApp plugin not available');\n"
            + "            return _P;\n"
            + "        }\n"
            + "\n"
            + "        window.WebAppLocalServer = {\n"
            + "            startupDidComplete(callback) {\n"
            + "                var P = getPlugin(); if (!P) return;\n"
            + "                P.startupDidComplete()\n"
            + "                .then(function() { if (callback) callback(); })\n"
            + "                .catch(function(error) { console.error('WebAppLocalServer.startupDidComplete() failed:', error); });\n"
            + "            },\n"
            + "\n"
            + "            checkForUpdates(callback) {\n"
            + "                var P = getPlugin(); if (!P) return;\n"
            + "                P.checkForUpdates()\n"
            + "                .then(function() { if (callback) callback(); })\n"
            + "                .catch(function(error) { console.error('WebAppLocalServer.checkForUpdates() failed:', error); });\n"
            + "            },\n"
            + "\n"
            + "            onNewVersionReady(callback) {\n"
            + "                var P = getPlugin(); if (!P) return;\n"
            + "                P.addListener('updateAvailable', function(event) { callback(event.version); });\n"
            + "            },\n"
            + "\n"
            + "            switchToPendingVersion(callback, errorCallback) {\n"
            + "                var P = getPlugin(); if (!P) return;\n"
            + "                P.reload()\n"
            + "                .then(function() { if (callback) callback(); })\n"
            + "                .catch(function(error) {\n"
            + "                    console.error('switchToPendingVersion failed:', error);\n"
            + "                    if (typeof errorCallback === 'function') errorCallback(error);\n"
            + "                });\n"
            + "            },\n"
            + "\n"
            + "            onError(callback) {\n"
            + "                var P = getPlugin(); if (!P) return;\n"
            + "                P.addListener('error', function(event) {\n"
            + "                    var error = new Error(event.message || 'Unknown CapacitorMeteorWebApp error');\n"
            + "                    callback(error);\n"
            + "                });\n"
            + "            },\n"
            + "\n"
            + "            localFileSystemUrl(_fileUrl) {\n"
            + "                throw new Error('Local filesystem URLs not supported by Capacitor');\n"
            + "            },\n"
            + "        };\n"
            + "    }\n"
            + "})();\n"
            + "</script>";

    private BundleOrganizer() {}

    static void organizeBundle(AssetBundle bundle, File targetDirectory) throws WebAppError {
        List<String> validationErrors = validateBundleOrganization(bundle);
        if (!validationErrors.isEmpty()) {
            throw new WebAppError(
                WebAppError.Type.UNSUITABLE_ASSET_BUNDLE,
                "Bundle validation failed: " + String.join("; ", validationErrors)
            );
        }

        if (!targetDirectory.exists() && !targetDirectory.mkdirs() && !targetDirectory.isDirectory()) {
            throw new WebAppError(
                WebAppError.Type.FILE_SYSTEM_ERROR,
                "Failed to create bundle target directory: " + targetDirectory.getAbsolutePath()
            );
        }

        List<String> skippedAssets = new ArrayList<>();

        for (Asset asset : bundle.getOwnAssets()) {
            organizeAsset(asset, targetDirectory, skippedAssets);
        }

        AssetBundle parentBundle = bundle.getParentAssetBundle();
        if (parentBundle != null) {
            for (Asset parentAsset : parentBundle.getOwnAssets()) {
                if (bundle.getOwnAssetsByUrlPath().containsKey(parentAsset.urlPath)) {
                    continue;
                }
                organizeAsset(parentAsset, targetDirectory, skippedAssets);
            }
        }

        if (!skippedAssets.isEmpty()) {
            logWarning("Bundle organization completed with " + skippedAssets.size()
                + " missing asset(s) | bundleVersion=" + bundle.getVersion()
                + " | skipped=" + skippedAssets);
        }
    }

    static List<String> validateBundleOrganization(AssetBundle bundle) {
        List<String> errors = new ArrayList<>();
        Set<String> normalizedUrlPaths = new HashSet<>();

        for (Asset asset : bundle.getOwnAssets()) {
            String normalizedUrlPath;
            try {
                normalizedUrlPath = normalizeUrlPath(asset.urlPath);
            } catch (IllegalArgumentException e) {
                errors.add("Invalid urlPath for asset '" + asset.urlPath + "': " + e.getMessage());
                continue;
            }

            if (!normalizedUrlPaths.add(normalizedUrlPath)) {
                errors.add("Duplicate normalized URL path: " + normalizedUrlPath);
            }

            try {
                normalizeRelativeFilePath(asset.filePath);
            } catch (IllegalArgumentException e) {
                errors.add("Invalid filePath for asset '" + asset.filePath + "': " + e.getMessage());
            }
        }

        return errors;
    }

    static File targetFileForAsset(Asset asset, File targetDirectory) {
        String normalizedUrlPath = normalizeUrlPath(asset.urlPath);
        return new File(targetDirectory, normalizedUrlPath);
    }

    static void cleanupOrganizedBundle(File targetDirectory) throws WebAppError {
        if (!targetDirectory.exists()) {
            return;
        }

        if (!FileOps.deleteRecursively(targetDirectory)) {
            throw new WebAppError(
                WebAppError.Type.FILE_SYSTEM_ERROR,
                "Failed to cleanup organized bundle at " + targetDirectory.getAbsolutePath()
            );
        }
    }

    private static void organizeAsset(Asset asset, File targetDirectory, List<String> skippedAssets)
        throws WebAppError {
        final File targetFile;
        try {
            targetFile = targetFileForAsset(asset, targetDirectory);
        } catch (IllegalArgumentException e) {
            throw new WebAppError(
                WebAppError.Type.UNSUITABLE_ASSET_BUNDLE,
                "Invalid asset URL path while organizing: " + asset.urlPath,
                e
            );
        }
        ensureTargetWithinDirectory(targetFile, targetDirectory);

        try {
            FileOps.ensureParentDirectory(targetFile);

            if (isIndexAsset(asset)) {
                organizeIndexHtml(asset, targetFile);
                return;
            }

            try (InputStream source = asset.bundle.openAsset(asset.filePath)) {
                FileOps.copy(source, targetFile);
            }
        } catch (FileNotFoundException e) {
            // index.html is handled by organizeIndexHtml() above which throws its
            // own WebAppError if the file is missing, so only non-index assets
            // reach this catch block.
            if (isSourceMapAsset(asset)) {
                return;
            }
            logWarning("Missing asset skipped during bundle organization"
                + " | urlPath=" + asset.urlPath
                + " | filePath=" + asset.filePath
                + " | bundleVersion=" + asset.bundle.getVersion());
            skippedAssets.add(asset.urlPath);
        } catch (IOException e) {
            throw new WebAppError(
                WebAppError.Type.FILE_SYSTEM_ERROR,
                "Failed to organize asset: " + asset.urlPath,
                e
            );
        }
    }

    private static void organizeIndexHtml(Asset asset, File targetFile) throws WebAppError {
        try (InputStream source = asset.bundle.openAsset(asset.filePath)) {
            String originalContent = readUtf8(source);
            String modifiedContent = injectShim(originalContent);
            try (InputStream output = new ByteArrayInputStream(modifiedContent.getBytes(StandardCharsets.UTF_8))) {
                FileOps.copy(output, targetFile);
            }
        } catch (IOException e) {
            throw new WebAppError(
                WebAppError.Type.FILE_SYSTEM_ERROR,
                "Failed to process index.html",
                e
            );
        }
    }

    private static String readUtf8(InputStream inputStream) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[4096];
        int read;
        while ((read = inputStream.read(buffer)) != -1) {
            output.write(buffer, 0, read);
        }
        return new String(output.toByteArray(), StandardCharsets.UTF_8);
    }

    private static String injectShim(String html) {
        String headTag = "</head>";
        String bodyTag = "</body>";

        int headIndex = indexOfIgnoreCase(html, headTag);
        if (headIndex >= 0) {
            return html.substring(0, headIndex) + SHIM_SCRIPT + "\n" + html.substring(headIndex);
        }

        int bodyIndex = indexOfIgnoreCase(html, bodyTag);
        if (bodyIndex >= 0) {
            return html.substring(0, bodyIndex) + SHIM_SCRIPT + "\n" + html.substring(bodyIndex);
        }

        return html + SHIM_SCRIPT;
    }

    private static int indexOfIgnoreCase(String source, String value) {
        return source.toLowerCase(Locale.US).indexOf(value.toLowerCase(Locale.US));
    }

    private static boolean isIndexAsset(Asset asset) {
        return "/".equals(asset.urlPath) || "/index.html".equals(asset.urlPath);
    }

    private static boolean isSourceMapAsset(Asset asset) {
        return asset.urlPath.endsWith(".map") || asset.filePath.endsWith(".map");
    }

    private static void logWarning(String message) {
        try {
            Log.w(LOG_TAG, message);
        } catch (RuntimeException ignored) {
            // Local JVM tests do not mock android.util.Log; keep behavior test-safe.
            System.err.println(LOG_TAG + " WARN: " + message);
        }
    }

    private static String normalizeUrlPath(String rawUrlPath) {
        if (rawUrlPath == null || rawUrlPath.isEmpty()) {
            throw new IllegalArgumentException("path is empty");
        }

        String withoutQuery = WebAppUtils.removeQueryStringFromUrlPath(rawUrlPath);
        String path = withoutQuery;
        if (path.startsWith("/")) {
            path = path.substring(1);
        }

        if (path.isEmpty()) {
            return "index.html";
        }

        return normalizePath(path);
    }

    private static String normalizeRelativeFilePath(String rawFilePath) {
        if (rawFilePath == null || rawFilePath.isEmpty()) {
            throw new IllegalArgumentException("path is empty");
        }

        if (rawFilePath.startsWith("/")) {
            throw new IllegalArgumentException("absolute paths are not allowed");
        }

        return normalizePath(rawFilePath);
    }

    private static String normalizePath(String path) {
        if (path.contains("\\")) {
            throw new IllegalArgumentException("backslashes are not allowed");
        }

        String[] segments = path.split("/");
        StringBuilder normalized = new StringBuilder();

        for (String segment : segments) {
            if (segment.isEmpty()) {
                throw new IllegalArgumentException("empty path segments are not allowed");
            }
            if (".".equals(segment) || "..".equals(segment)) {
                throw new IllegalArgumentException("path traversal segments are not allowed");
            }

            if (normalized.length() > 0) {
                normalized.append('/');
            }
            normalized.append(segment);
        }

        return normalized.toString();
    }

    private static void ensureTargetWithinDirectory(File targetFile, File targetDirectory) throws WebAppError {
        try {
            String targetCanonical = targetFile.getCanonicalPath();
            String rootCanonical = targetDirectory.getCanonicalPath();

            if (!targetCanonical.equals(rootCanonical) && !targetCanonical.startsWith(rootCanonical + File.separator)) {
                throw new WebAppError(
                    WebAppError.Type.UNSUITABLE_ASSET_BUNDLE,
                    "Asset target escapes serving directory: " + targetFile.getPath()
                );
            }
        } catch (IOException e) {
            throw new WebAppError(
                WebAppError.Type.FILE_SYSTEM_ERROR,
                "Failed to validate serving target path",
                e
            );
        }
    }
}

File: npm-packages/meteor-capacitor/ios/Sources/MeteorjsCapacitor/Bundle/BundleOrganizer.swift
// Vendored from https://github.com/banjerluke/capacitor-meteor-webapp
// Copyright 2025 Luke Abbott. MIT license. See LICENSES/banjerluke-capacitor-meteor-webapp.txt.

//
// BundleOrganizer.swift
//
// Handles file organization logic for bundles, including URL path mapping
// and directory structure creation for Meteor webapp assets.
//

import Foundation
import os.log

/// Handles file organization logic for bundles, including URL path mapping and directory structure creation
public class BundleOrganizer {

    private static let logger = os.Logger(
        subsystem: "com.meteor.webapp", category: "BundleOrganizer")

    private struct PathValidationError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    /// Organizes files in a bundle directory according to their URL mappings
    /// - Parameters:
    ///   - bundle: The asset bundle to organize
    ///   - targetDirectory: The directory where files should be organized
    /// - Throws: WebAppError if organization fails
    static func organizeBundle(_ bundle: AssetBundle, in targetDirectory: URL) throws {
        // Validate bundle before any file operations to prevent path traversal
        // and duplicate URL attacks from malicious manifests
        let validationErrors = validateBundleOrganization(bundle)
        if !validationErrors.isEmpty {
            throw WebAppError.unsuitableAssetBundle(
                reason: "Bundle validation failed: \(validationErrors.joined(separator: "; "))",
                underlyingError: nil)
        }

        let fileManager = FileManager.default

        // Create target directory if it doesn't exist
        try fileManager.createDirectory(
            at: targetDirectory, withIntermediateDirectories: true, attributes: nil)

        var skippedAssets: [String] = []

        // Organize own assets
        for asset in bundle.ownAssets {
            try organizeAsset(asset, in: targetDirectory, fileManager: fileManager, skippedAssets: &skippedAssets)
        }

        // Also organize parent assets that this bundle inherits but doesn't override
        if let parentBundle = bundle.parentAssetBundle {
            for parentAsset in parentBundle.ownAssets {
                // Only organize parent assets that we don't have in our own assets
                if bundle.ownAssetsByURLPath[parentAsset.urlPath] == nil {
                    try organizeAsset(parentAsset, in: targetDirectory, fileManager: fileManager, skippedAssets: &skippedAssets)
                }
            }
        }

        if !skippedAssets.isEmpty {
            logger.warning("Bundle organization completed with \(skippedAssets.count) missing asset(s) | bundleVersion=\(bundle.version) | skipped=\(skippedAssets)")
        }
    }

    /// Organizes a single asset according to its URL path mapping
    /// - Parameters:
    ///   - asset: The asset to organize
    ///   - targetDirectory: The target directory
    ///   - fileManager: File manager instance
    /// - Throws: WebAppError if organization fails
    private static func organizeAsset(
        _ asset: Asset, in targetDirectory: URL, fileManager: FileManager,
        skippedAssets: inout [String]
    ) throws {
        let sourceURL = asset.fileURL
        let targetURL = targetURLForAsset(asset, in: targetDirectory)

        // Ensure the target directory structure exists
        let targetDirectoryURL = targetURL.deletingLastPathComponent()
        try fileManager.createDirectory(
            at: targetDirectoryURL, withIntermediateDirectories: true, attributes: nil)

        // Check if source file exists
        guard fileManager.fileExists(atPath: sourceURL.path) else {
            if asset.urlPath.hasSuffix(".map") || asset.fileURL.pathExtension == "map" {
                // Skip missing source maps silently
                return
            }
            if asset.urlPath == "/" || asset.urlPath == "/index.html" {
                logger.error("Source file missing for critical asset: \(asset.urlPath), Expected path: \(sourceURL.path)")
                throw WebAppError.fileSystemError(
                    reason: "Source file does not exist: \(sourceURL.path)", underlyingError: nil)
            }
            logger.warning("Missing asset skipped during bundle organization | urlPath=\(asset.urlPath) | filePath=\(asset.filePath) | bundleVersion=\(asset.bundle.version)")
            skippedAssets.append(asset.urlPath)
            return
        }

        // If target already exists, remove it first
        if fileManager.fileExists(atPath: targetURL.path) {
            try fileManager.removeItem(at: targetURL)
        }

        if asset.urlPath == "/" || asset.urlPath == "/index.html" {
            // Special handling for index.html - inject WebAppLocalServer shim
            try organizeIndexHtml(sourceURL: sourceURL, targetURL: targetURL)
        } else {
            // Try to create hard link first (for efficiency), fall back to copy
            do {
                try fileManager.linkItem(at: sourceURL, to: targetURL)
            } catch {
                // Hard link failed, try copying instead
                do {
                    try fileManager.copyItem(at: sourceURL, to: targetURL)
                } catch {
                    throw WebAppError.fileSystemError(
                        reason: "Failed to organize asset \(asset.urlPath)", underlyingError: error)
                }
            }
        }
    }

    /// Special handling for index.html files to inject WebAppLocalServer shim
    /// - Parameters:
    ///   - sourceURL: Source index.html file
    ///   - targetURL: Target location for the modified index.html
    /// - Throws: WebAppError if processing fails
    private static func organizeIndexHtml(sourceURL: URL, targetURL: URL) throws {
        // Read the original HTML content
        let originalContent: String
        do {
            originalContent = try String(contentsOf: sourceURL, encoding: .utf8)
        } catch {
            throw WebAppError.fileSystemError(
                reason: "Failed to read index.html content", underlyingError: error)
        }

        // WebAppLocalServer compatibility shim for Capacitor
        // Provides the same API as cordova-plugin-meteor-webapp
        let shimScript = """
            <script>
            (function() {
                if (window.WebAppLocalServer) return;

                setupWebAppLocalServer();

                function setupWebAppLocalServer() {
                    // Resolve plugin lazily — the Capacitor bridge may not have
                    // registered plugins yet when the shim first initializes
                    var _P;
                    function getPlugin() {
                        if (!_P) _P = ((window.Capacitor || {}).Plugins || {}).CapacitorMeteorWebApp;
                        if (!_P) console.warn('WebAppLocalServer shim: CapacitorMeteorWebApp plugin not available');
                        return _P;
                    }

                    window.WebAppLocalServer = {
                        startupDidComplete(callback) {
                            var P = getPlugin(); if (!P) return;
                            P.startupDidComplete()
                            .then(function() { if (callback) callback(); })
                            .catch(function(error) { console.error('WebAppLocalServer.startupDidComplete() failed:', error); });
                        },

                        checkForUpdates(callback) {
                            var P = getPlugin(); if (!P) return;
                            P.checkForUpdates()
                            .then(function() { if (callback) callback(); })
                            .catch(function(error) { console.error('WebAppLocalServer.checkForUpdates() failed:', error); });
                        },

                        onNewVersionReady(callback) {
                            var P = getPlugin(); if (!P) return;
                            P.addListener('updateAvailable', function(event) { callback(event.version); });
                        },

                        switchToPendingVersion(callback, errorCallback) {
                            var P = getPlugin(); if (!P) return;
                            P.reload()
                            .then(function() { if (callback) callback(); })
                            .catch(function(error) {
                                console.error('switchToPendingVersion failed:', error);
                                if (typeof errorCallback === 'function') errorCallback(error);
                            });
                        },

                        onError(callback) {
                            var P = getPlugin(); if (!P) return;
                            P.addListener('error', function(event) {
                                var error = new Error(event.message || 'Unknown CapacitorMeteorWebApp error');
                                callback(error);
                            });
                        },

                        localFileSystemUrl(_fileUrl) {
                            throw new Error('Local filesystem URLs not supported by Capacitor');
                        },
                    };
                }
            })();
            </script>
        """

        // Inject the shim before closing </head> tag, or before </body> if no head
        let modifiedContent: String
        if let headCloseRange = originalContent.range(of: "</head>", options: .caseInsensitive) {
            modifiedContent = originalContent.replacingCharacters(in: headCloseRange, with: shimScript + "\n</head>")
        } else if let bodyCloseRange = originalContent.range(of: "</body>", options: .caseInsensitive) {
            modifiedContent = originalContent.replacingCharacters(in: bodyCloseRange, with: shimScript + "\n</body>")
        } else {
            // Just append to the end if we can't find head or body tags
            modifiedContent = originalContent + shimScript
        }

        do {
            try modifiedContent.write(to: targetURL, atomically: true, encoding: .utf8)
        } catch {
            throw WebAppError.fileSystemError(reason: "Failed to write modified index.html", underlyingError: error)
        }
    }

    /// Calculates the target URL for an asset based on its URL path mapping
    /// - Parameters:
    ///   - asset: The asset
    ///   - targetDirectory: The target directory
    /// - Returns: The URL where the asset should be placed
    static func targetURLForAsset(_ asset: Asset, in targetDirectory: URL) -> URL {
        // Remove leading slash from URL path to make it relative
        var relativePath = asset.urlPath
        if relativePath.hasPrefix("/") {
            relativePath = String(relativePath.dropFirst())
        }

        // Handle root path (/) -> index.html
        if relativePath.isEmpty {
            relativePath = "index.html"
        }

        return targetDirectory.appendingPathComponent(relativePath)
    }

    /// Validates that all assets in a bundle can be properly organized
    /// - Parameter bundle: The bundle to validate
    /// - Returns: Array of validation errors (empty if valid)
    static func validateBundleOrganization(_ bundle: AssetBundle) -> [String] {
        var errors: [String] = []
        var normalizedURLPaths = Set<String>()

        for asset in bundle.ownAssets {
            let normalizedURLPath: String
            do {
                normalizedURLPath = try normalizeURLPath(asset.urlPath)
            } catch {
                errors.append(
                    "Invalid urlPath for asset '\(asset.urlPath)': \(error.localizedDescription)")
                continue
            }

            if !normalizedURLPaths.insert(normalizedURLPath).inserted {
                errors.append("Duplicate normalized URL path: \(normalizedURLPath)")
            }

            do {
                _ = try normalizeRelativeFilePath(asset.filePath)
            } catch {
                errors.append(
                    "Invalid filePath for asset '\(asset.filePath)': \(error.localizedDescription)")
            }
        }

        return errors
    }

    private static func normalizeURLPath(_ rawURLPath: String) throws -> String {
        if rawURLPath.isEmpty {
            throw PathValidationError(message: "path is empty")
        }

        var path = URLPathByRemovingQueryString(rawURLPath)
        if path.hasPrefix("/") {
            path.removeFirst()
        }

        if path.isEmpty {
            return "index.html"
        }

        return try normalizePath(path)
    }

    private static func normalizeRelativeFilePath(_ rawFilePath: String) throws -> String {
        if rawFilePath.isEmpty {
            throw PathValidationError(message: "path is empty")
        }

        if rawFilePath.hasPrefix("/") {
            throw PathValidationError(message: "absolute paths are not allowed")
        }

        return try normalizePath(rawFilePath)
    }

    private static func normalizePath(_ path: String) throws -> String {
        if path.contains("\\") {
            throw PathValidationError(message: "backslashes are not allowed")
        }

        let segments = path.split(separator: "/", omittingEmptySubsequences: false)
        var normalizedSegments: [Substring] = []

        for segment in segments {
            if segment.isEmpty {
                throw PathValidationError(message: "empty path segments are not allowed")
            }

            if segment == "." || segment == ".." {
                throw PathValidationError(message: "path traversal segments are not allowed")
            }

            normalizedSegments.append(segment)
        }

        return normalizedSegments.joined(separator: "/")
    }

    /// Removes organized files from a target directory
    /// - Parameter targetDirectory: Directory to clean up
    /// - Throws: WebAppError if cleanup fails
    public static func cleanupOrganizedBundle(at targetDirectory: URL) throws {
        let fileManager = FileManager.default

        guard fileManager.fileExists(atPath: targetDirectory.path) else {
            return  // Nothing to clean up
        }

        do {
            try fileManager.removeItem(at: targetDirectory)
        } catch {
            throw WebAppError.fileSystemError(
                reason: "Failed to cleanup bundle directory", underlyingError: error)
        }
    }
}
